cairo/ethereum/utils/numeric.cairo

```diff
@@ -336,3 +383,179 @@ func Bytes32_from_be_bytes{range_check_ptr, bitwise_ptr: BitwiseBuiltin*}(bytes:
     tempvar res_bytes32 = Bytes32(new Bytes32Struct(res_reversed.low, res_reversed.high));
     return res_bytes32;
 }
+
+func is_U384_zero(num: U384) -> felt {
+    if (num.value.d0 == 0 and num.value.d1 == 0 and num.value.d2 == 0 and num.value.d3 == 0) {
+        return 1;
+    }
+    return 0;
+}
+
+func is_U384_one(num: U384) -> felt {
+    if (num.value.d0 == 1 and num.value.d1 == 0 and num.value.d2 == 0 and num.value.d3 == 0) {
+        return 1;
+    }
+    return 0;
+}
+
+func U384_from_be_bytes{range_check_ptr, bitwise_ptr: BitwiseBuiltin*}(bytes: Bytes) -> U384 {
+    alloc_locals;
+
+    // Verify that the bytes length is not greater than 48 (384 bits)
+    with_attr error_message("ValueError") {
+        assert [range_check_ptr] = 48 - bytes.value.len;
+        let range_check_ptr = range_check_ptr + 1;
+    }
+
+    // For empty input, return 0
+    if (bytes.value.len == 0) {
+        tempvar res = U384(new UInt384(d0=0, d1=0, d2=0, d3=0));
+        return res;
+    }
+
+    // Calculate how many bytes go into each 96-bit limb
+    // Each limb can hold up to 12 bytes (96 bits)
+    let d0_len = min(12, bytes.value.len);
+    let max_d1_len = max(0, bytes.value.len - 12);
+    let d1_len = min(12, max_d1_len);
+    let max_d2_len = max(0, bytes.value.len - 24);
+    let d2_len = min(12, max_d2_len);
+    let max_d3_len = max(0, bytes.value.len - 36);
+    let d3_len = min(12, max_d3_len);
+
+    // Extract bytes for each limb
+    // Note: We need to process from the least significant (rightmost) bytes to the most significant
+    let d0_start = bytes.value.len - d0_len;
+    let d1_start = max(0, bytes.value.len - d0_len - d1_len);
+    let d2_start = max(0, bytes.value.len - d0_len - d1_len - d2_len);
+    let d3_start = 0;  // Most significant bytes start at index 0
+
+    // Create byte slices for each limb
+    tempvar d0_bytes = Bytes(new BytesStruct(data=bytes.value.data + d0_start, len=d0_len));
+    tempvar d1_bytes = Bytes(new BytesStruct(data=bytes.value.data + d1_start, len=d1_len));
+    tempvar d2_bytes = Bytes(new BytesStruct(data=bytes.value.data + d2_start, len=d2_len));
+    tempvar d3_bytes = Bytes(new BytesStruct(data=bytes.value.data + d3_start, len=d3_len));
+
+    // Convert each byte slice to a felt
+    let d0 = bytes_to_felt(d0_bytes.value.len, d0_bytes.value.data);
+    let d1 = bytes_to_felt(d1_bytes.value.len, d1_bytes.value.data);
+    let d2 = bytes_to_felt(d2_bytes.value.len, d2_bytes.value.data);
+    let d3 = bytes_to_felt(d3_bytes.value.len, d3_bytes.value.data);
+
+    // Create the U384 value
+    tempvar res = U384(new UInt384(d0=d0, d1=d1, d2=d2, d3=d3));
+    return res;
+}
+
+func U384__eq__(a: U384, b: U384) -> bool {
```

Comment:
there exists a uint384_eq i think from corelib

Reply:
there is a uint384_eq in cairo_ec indeed, will leverage it